adr/ADR-31.md

```diff
@@ -115,6 +115,30 @@ After the batch is sent a zero length payload message will be sent with the `Nat
 
 When requests are made against servers that do not support `batch` the first response will be received and nothing will follow. Old servers can be detected by the absence of the `Nats-Num-Pending` header in the first reply.
 
+There are 4 viable API calls for a batch. All require a batch amount greater than 0 and a subject which may include a wildcard.
+They also require a start sequence -or- a start time. 
+There is a server issue under consideration requesting that if neither start sequence nor a start time is supplied that it defaults to start sequence of 1.
+For now the client can optionally provide the 2 additional calls which provide the start sequence of 1 for the user. 
+
+1. get up to batch number of messages >= than sequence 1
+    * API: `batch: number, subject: string`
+    * Request: `{"batch":3,"seq":1,"next_by_subj":"foo.>"}`
+1. get up to batch number of messages >= than sequence that has specified subject
+    * API: `batch: number, sequence: number, subject: string`
+    * Request: `{"batch":3,"seq":4,"next_by_subj":"foo.>"}`
+1. gets up to batch number of messages >= than start time that has specified subject 
+    * API: `batch: number, start time: time, subject: string`
+    * Request: `{"batch":3,"start_time":"2024-11-04T23:45:02.060192000Z","next_by_subj":"foo.>"}`
+1. get up to batch number of messages >= than sequence 1, limited by max bytes
+    * API: `batch: number, max_bytes: number, sequence: number, subject: string`
+    * Request: `{"batch":3,"max_bytes":2002,"seq":1,"next_by_subj":"foo.>"}`
+1. get up to batch number of messages >= than sequence that has specified subject, limited by max bytes
+    * API: `batch: number, max_bytes: number, sequence: number, subject: string`
+    * Request: `{"batch":3,"max_bytes":2002,"seq":4,"next_by_subj":"foo.>"}`
+1. gets up to batch number of messages >= than start time that has specified subject, limited by max bytes
+    * API: `batch: number, max_bytes: number, start time: time, subject: string`
```

Comment:
`start_time` is not supported in batch requests (possibly should be)

Reply:
I have unit tests that say otherwise.

Reply:
Check: https://github.com/nats-io/nats.js/blob/4fb61dc8ad21b7467569758fa10cd6351bda7561/jetstream/tests/jsm_direct_test.ts#L327

Reply:
The start time is ignored.

Reply:
I have unit tests for all of these and they all work as expected.
1. batch only
1. batch with starting sequence
1. batch with start time
1. batch with max bytes
1. batch with max bytes and starting sequence
1. batch with max bytes and start time

For multi last I have these tests
1. just subjects
1. subjects with up_to_seq
1. subjects with up_to_time

Here is the server validation code (Stream.go line 4266):
```
// Check we don't have conflicting options set.
// We do not allow batch mode for lastFor requests.
if (req.Seq > 0 && req.LastFor != _EMPTY_) ||
	(req.Seq > 0 && req.StartTime != nil) ||
	(req.StartTime != nil && req.LastFor != _EMPTY_) ||
	(req.LastFor != _EMPTY_ && req.NextFor != _EMPTY_) ||
	(req.LastFor != _EMPTY_ && req.Batch > 0) ||
	(req.LastFor != _EMPTY_ && len(req.MultiLastFor) > 0) ||
	(req.NextFor != _EMPTY_ && len(req.MultiLastFor) > 0) ||
	(req.UpToSeq > 0 && req.UpToTime != nil) {
	hdr := []byte("NATS/1.0 408 Bad Request\r\n\r\n")
	mset.outq.send(newJSPubMsg(reply, _EMPTY_, _EMPTY_, hdr, nil, nil, 0))
	return
}
```

Relevant notes on that code
* seq and last for cannot both be present
* seq and start time cannot both be present
* up_to_seq and up_to_time cannot both be present

Reply:
It appears that batch works to limit the multi_for response, so I have to add these variants in